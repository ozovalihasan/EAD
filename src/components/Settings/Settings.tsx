import useCustomizationStore from '@/zustandStore/customizationStore';
import { memo } from 'react';
import Select from 'react-select'
import { availableColors, availableColorsType, setColorVariants } from '@/components';

export const Settings = memo(() => {
  
  const locationSidebar = useCustomizationStore(store => store.locationSidebar)
  const showTextOnEdges = useCustomizationStore(store => store.showTextOnEdges)
  const toggleLocationSidebar = useCustomizationStore(store => store.toggleLocationSidebar)
  const toggleTextMode = useCustomizationStore(store => store.toggleTextMode)

  return (
    <details open className='mt-8 [&>summary>span:nth-child(1)]:open:hidden [&>summary>span:nth-child(2)]:open:inline '>
      <summary className='btn-third'>
        <span>Show menu</span>
        <span className='hidden'>Hide menu</span> 
      </summary>
      <div className='mt-8 flex-col flex'>
        <label className='flex mb-5 cursor-pointer'>
          <input className='mr-6' type="checkbox" checked={ showTextOnEdges} onChange={toggleTextMode}/>
          Show Association Names
        </label>
        <label className='flex mb-5 cursor-pointer'>
          <input className='mr-6' type="checkbox" checked={ locationSidebar === "right"} onChange={toggleLocationSidebar}/>
          Show the sidebar at the right of the window
        </label>
        <Select 
          className="w-full capitalize" 
          onChange={(selectedOption) => setColorVariants((selectedOption!).value as availableColorsType)} 
          placeholder="Select main color..."
          options={
            availableColors.map((color: string) => ({ value: color, label: color }))
          } 
        />
      </div>
    </details>
  )
})