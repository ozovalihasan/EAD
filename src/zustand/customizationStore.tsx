import create from 'zustand';
import { devtools } from 'zustand/middleware'


type CustomizationStoreState = {
  locationSidebar: string | null;
  widthSidebar: number | null;
  showTextOnEdges: boolean;
  toggleLocationSidebar: () => void;
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => void;
  toggleTextMode: () => void;
}

if(!localStorage.locationSidebar){
  let locationSidebar = "left"
  localStorage.setItem("locationSidebar", JSON.stringify(locationSidebar))
}

if(!localStorage.widthSidebar){
  let widthSidebar = Math.floor(window.innerWidth * 0.2)
  localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar))
}

const useCustomizationStore = create(devtools<CustomizationStoreState>((set, get) => ({
  locationSidebar: JSON.parse(localStorage.locationSidebar),
  widthSidebar: JSON.parse(localStorage.widthSidebar),
  showTextOnEdges: false,
  toggleLocationSidebar: () => {
    let location = JSON.parse(localStorage.locationSidebar)
    
    if (location === "left"){
      location = "right"
    } else {
      location = "left"
    }
    localStorage.setItem("locationSidebar", JSON.stringify( location) )  

    let widthSidebar = window.innerWidth - (get().widthSidebar as number); 
    localStorage.setItem("widthSidebar", JSON.stringify( widthSidebar) )  

    set({
      locationSidebar: location,
      widthSidebar: widthSidebar
    })
  },
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => {
    const widthSidebar = e.clientX;
    localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar) );
    set({
      widthSidebar: widthSidebar
    })
  },
  toggleTextMode: () => {
    set({
        showTextOnEdges: !get().showTextOnEdges
    })
  }
})))

export default useCustomizationStore;