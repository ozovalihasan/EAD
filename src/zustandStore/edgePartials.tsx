export type HasOneEdgePartialType =  {
  type: "hasOne",
  label: "has one"
}

export const hasOneEdgePartial: HasOneEdgePartialType= {
  type: "hasOne",
  label: "has one"
}

export type HasManyEdgePartialType = {
  type: "hasMany",
  label: "has many"
}

export const hasManyEdgePartial: HasManyEdgePartialType = {
  type: "hasMany",
  label: "has many"
}

export type ThroughEdgePartialType = {
  type: "through",
  label: "through"
}

export const throughEdgePartial: ThroughEdgePartialType = {
  type: "through",
  label: "through"
}