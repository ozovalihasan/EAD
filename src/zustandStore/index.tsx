export * from './edgePartials';
export * from './nodePartials';
export * from './store';
