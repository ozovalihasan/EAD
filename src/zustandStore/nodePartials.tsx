export const entityNodePartial: {
  type: "entity"
} = {
  type: "entity"
}