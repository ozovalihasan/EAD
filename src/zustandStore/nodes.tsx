import { Node } from "react-flow-renderer";

const initialNodes: Node[] = [
  {
    "id": "4",
    "type": "entity",
    "position": {
        "x": 465,
        "y": 150
    },
    "data": {
        "tableId": "1",
        "name": "Physician"
    },
    "width": 138,
    "height": 45
  },
  {
      "id": "5",
      "type": "entity",
      "position": {
          "x": 585,
          "y": 255
      },
      "data": {
          "tableId": "2",
          "name": "Appointment"
      },
      "width": 138,
      "height": 45,
      "selected": false,
      "positionAbsolute": {
          "x": 585,
          "y": 255
      },
      "dragging": false
  },
  {
      "id": "6",
      "type": "entity",
      "position": {
          "x": 720,
          "y": 150
      },
      "data": {
          "tableId": "3",
          "name": "Patient"
      },
      "width": 138,
      "height": 45,
      "selected": false,
      "positionAbsolute": {
          "x": 720,
          "y": 150
      },
      "dragging": false
  }
];

export default initialNodes