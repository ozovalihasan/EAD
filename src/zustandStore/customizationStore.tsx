import { availableColors, availableColorsType, setColorVariants } from '@/helpers';
import create from 'zustand';
import { devtools } from 'zustand/middleware'


export enum LocationSidebar {
  Left = "left",
  Right = "right"
}

export interface CustomizationStoreState {
  locationSidebar: LocationSidebar;
  widthSidebar: number;
  showTextOnEdges: boolean;
  sidebarVisible: boolean;
  navbarVisible: boolean;
  mainColor: availableColorsType;
  darkModeActive: boolean;
  changeMainColor: (color: availableColorsType) => void;
  toggleLocationSidebar: () => void;
  toggleSidebarVisibility: () => void;
  toggleNavbarVisibility: () => void;
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => void;
  toggleTextMode: () => void;
  toggleDarkMode: () => void;
}

export const checkLocalStorage = () => {

  if(!localStorage.locationSidebar){
    const locationSidebar = LocationSidebar.Left
    localStorage.setItem("locationSidebar", JSON.stringify(locationSidebar))
  }

  if(!localStorage.widthSidebar){
    const widthSidebar = Math.floor(window.innerWidth * 0.2)
    localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar))
  }

  if(!localStorage.mainColor){
    localStorage.setItem("mainColor", JSON.stringify(availableColors[0]))
  }

  if(!localStorage.darkModeActive){
    localStorage.setItem("darkModeActive", JSON.stringify(false))
  }else{
    if (JSON.parse(localStorage.darkModeActive as string)){
      const rootElement = document.querySelector("div#root")!
      rootElement.classList.add("dark");
    }
  }
}

checkLocalStorage();

export const useCustomizationStore = create(devtools<CustomizationStoreState>((set, get) => ({
  locationSidebar: JSON.parse(localStorage.locationSidebar as string) as LocationSidebar,
  widthSidebar: JSON.parse(localStorage.widthSidebar as string) as number,
  showTextOnEdges: false,
  sidebarVisible: true,
  navbarVisible: true,
  mainColor: JSON.parse(localStorage.mainColor as string) as availableColorsType,
  darkModeActive: JSON.parse(localStorage.darkModeActive as string) as boolean, 
  changeMainColor: (color: availableColorsType) => {
    setColorVariants(color)

    set({
      mainColor: color,
    })
  },
  toggleLocationSidebar: () => {
    let location = get().locationSidebar
    
    if (location === LocationSidebar.Left){
      location = LocationSidebar.Right
    } else {
      location = LocationSidebar.Left
    }

    localStorage.setItem("locationSidebar", JSON.stringify( location) )  

    set({
      locationSidebar: location,
    })
  },
  toggleSidebarVisibility:() => {
    set({
      sidebarVisible: !get().sidebarVisible,
    })
  },
  toggleNavbarVisibility:() => {
    set({
      navbarVisible: !get().navbarVisible,
    })
  },
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => {
    let widthSidebar

    if (get().locationSidebar === LocationSidebar.Left){
      widthSidebar = e.clientX; 
    } else {
      widthSidebar = window.innerWidth - e.clientX; 
    }

    localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar) );
    set({
      widthSidebar: widthSidebar,
      sidebarVisible: true,
    })
  },
  toggleTextMode: () => {
    set({
        showTextOnEdges: !get().showTextOnEdges
    })
  },
  toggleDarkMode: () => {
    const rootElement = document.querySelector("div#root")!
    rootElement.classList.toggle("dark");

    localStorage.setItem("darkModeActive", JSON.stringify(!get().darkModeActive))

    set({
      darkModeActive: !get().darkModeActive
    })
  }
})));