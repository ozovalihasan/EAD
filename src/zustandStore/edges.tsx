import { CustomEdgeType } from "@/zustandStore/store";
import { hasManyEdgePartial, hasOneEdgePartial, throughEdgePartial } from "@/zustandStore/edgePartials";

const initialEdges: CustomEdgeType[] = [
  {
    "id": "7",
    "source": "4",
    "target": "5",
    "sourceHandle": "bottom2",
    "targetHandle": "top",
    "selected": false,
    ...hasManyEdgePartial,
    data: {
      optional: true
    }
  },
  {
    "id": "8",
    "source": "6",
    "target": "5",
    "sourceHandle": "bottom1",
    "targetHandle": "top",
    "selected": false,
    ...hasOneEdgePartial,
    data: {
      optional: false
    }
  },
  {
    "id": "9",
    "source": "4",
    "target": "6",
    "sourceHandle": "bottom3",
    "targetHandle": "top",
    "selected": false,
    ...throughEdgePartial,
    "data": {
        "throughNodeId": "5"
    }
  }  
];

export default initialEdges
