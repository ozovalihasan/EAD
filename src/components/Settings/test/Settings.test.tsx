import { Settings } from '../Settings';
import { render, screen, renderHook, fireEvent } from "@testing-library/react";
import { LocationSidebar, useCustomizationStore } from '@/zustandStore/customizationStore';


let renderReadyComponent: JSX.Element;

beforeEach(() => {
    
useCustomizationStore.setState({ 
    locationSidebar: LocationSidebar.Right,
    showTextOnEdges: true,
    toggleLocationSidebar: jest.fn(),
    toggleTextMode: jest.fn(),
    toggleDarkMode: jest.fn(),
  })
    
  renderReadyComponent = (
    <Settings />
  );
});

describe('<Settings />', () => {

  describe("if the menu is not open",() => {
    it('renders a checkbox to change the location of sidebar', () => {

      render(renderReadyComponent );

      const buttonEl = screen.getByText(/Settings/i);
      let labelElement = screen.queryByText(/Show Association Names/i)
      
      expect(labelElement).toBeNull();
      expect(buttonEl).toBeInTheDocument();
      
      fireEvent.click(buttonEl)

      labelElement = screen.queryByText(/Show Association Names/i)

      expect(labelElement).toBeInTheDocument();
    });
    it('renders correctly', () => {
      const renderedContainer = render(renderReadyComponent );

      expect(renderedContainer).toMatchSnapshot();
    });
  })
  
  describe("if the menu is open",() => {
    beforeEach(() => {
      render(renderReadyComponent );
      const buttonEl = screen.getByText(/Settings/i);
      
      expect(buttonEl).toBeInTheDocument();

      fireEvent.click(buttonEl)
    });
    
    it('renders a checkbox to change the location of sidebar', () => {
      render(renderReadyComponent );
      const { result } = renderHook(() => useCustomizationStore());

      const labelElement = screen.getByText(/Show the sidebar at the right of the window/i)
      const checkboxElement = labelElement.getElementsByTagName("input")[0]

      expect(checkboxElement.checked).toBe(true)
      
      fireEvent.click(checkboxElement)

      expect(result.current.toggleLocationSidebar).toHaveBeenCalledTimes(1);
    });

    it('renders a checkbox to change the visibility of edge names', () => {
      render(renderReadyComponent );
      const { result } = renderHook(() => useCustomizationStore());

      const labelElement = screen.getByText(/Show Association Names/i)
      const checkboxElement = labelElement.getElementsByTagName("input")[0]

      expect(checkboxElement.checked).toBe(true)
      
      fireEvent.click(checkboxElement)

      expect(result.current.toggleTextMode).toHaveBeenCalledTimes(1);
    });

    it('renders a button to change the dark mode', () => {

      render(renderReadyComponent );
      const { result } = renderHook(() => useCustomizationStore());
      const buttonEl = screen.getByText(/Dark Mode/i);
      
      expect(buttonEl).toBeInTheDocument();

      fireEvent.click(buttonEl)

      expect(result.current.toggleDarkMode).toHaveBeenCalledTimes(1);
    });

    it('renders the MainColorDropdown component', () => {
      render(renderReadyComponent );

      expect(screen.getByText(/MockMainColorDropdown/i)).toBeInTheDocument();
    });

    it('hides the menu when the mouse leaves the menu element', () => {
      render(renderReadyComponent );

      let labelElement = screen.queryByText(/Show Association Names/i)

      expect(labelElement).toBeInTheDocument();

      const menuEl = screen.getByText(/MockMainColorDropdown/i).parentNode as HTMLElement
      
      fireEvent.mouseLeave(menuEl)

      labelElement = screen.queryByText(/Show Association Names/i)

      expect(labelElement).toBeNull();
    });

    it('renders correctly', () => {
      const renderedContainer = render(renderReadyComponent );

      expect(renderedContainer).toMatchSnapshot();
    });
  });
});