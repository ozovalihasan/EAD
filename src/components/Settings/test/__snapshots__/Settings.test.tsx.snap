// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> if the menu is not open renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
      >
        <button
          class="btn-third"
          title="Settings"
        >
          Settings
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
    >
      <button
        class="btn-third"
        title="Settings"
      >
        Settings
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Settings /> if the menu is open renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
      >
        <button
          class="btn-third"
          title="Settings"
        >
          Settings
        </button>
        <div
          class="absolute right-0 top-0 flex-col flex bg-zinc-50 dark:bg-zinc-700 border border-solid general-border p-2 z-50 rounded-md "
        >
          <div
            class="text-center my-4"
          >
            Settings
          </div>
          <label
            class="flex mb-5 cursor-pointer"
          >
            <input
              checked=""
              class="mr-6 bg-first-100"
              type="checkbox"
            />
            Show Association Names
          </label>
          <label
            class="flex mb-5 cursor-pointer"
          >
            <input
              checked=""
              class="mr-6"
              type="checkbox"
            />
            Show the sidebar at the right of the window
          </label>
          <button
            class="btn-second rounded-md my-2 p-2 w-full"
          >
            Dark Mode
          </button>
          <div>
            MockMainColorDropdown
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="relative"
      >
        <button
          class="btn-third"
          title="Settings"
        >
          Settings
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
    >
      <button
        class="btn-third"
        title="Settings"
      >
        Settings
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
